src/QirRuntime/public/IQuantumApi.hpp

```diff
@@ -0,0 +1,146 @@
+        virtual void Unitary(long numTargets, double** unitary, Qubit targets[]) = 0;
+
+        virtual void Reset(Qubit target) = 0;
+
+        virtual void R(PauliId axis, Qubit target, double theta) = 0;
+        virtual void RFraction(PauliId axis, Qubit target, long numerator, int power) = 0;
+        virtual void R1(Qubit target, double theta) = 0;
+        virtual void R1Fraction(Qubit target, long numerator, int power) = 0;
+        virtual void Exp(long numTargets, PauliId paulis[], Qubit targets[], double theta) = 0;
+        virtual void ExpFraction(long numTargets, PauliId paulis[], Qubit targets[], long numerator, int power) = 0;
+
+        // Multicontrolled operations
+        virtual void ControlledX(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledY(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledZ(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledH(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledS(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledT(long numControls, Qubit controls[], Qubit target) = 0;
+        virtual void ControlledSWAP(long numControls, Qubit controls[], Qubit target1, Qubit target2) = 0;
+        virtual void ControlledClifford(
+            long numControls,
+            Qubit controls[],
+            CliffordId cliffordId,
+            PauliId pauli,
+            Qubit target) = 0;
+        virtual void ControlledUnitary(
+            long numControls,
+            Qubit controls[],
+            long numTargets,
+            double** unitary,
+            Qubit targets[]) = 0;
+
+        virtual void ControlledR(long numControls, Qubit controls[], PauliId axis, Qubit target, double theta) = 0;
+        virtual void ControlledRFraction(
+            long numControls,
+            Qubit controls[],
+            PauliId axis,
+            Qubit target,
+            long numerator,
+            int power) = 0;
+        virtual void ControlledR1(long numControls, Qubit controls[], Qubit target, double theta) = 0;
+        virtual void ControlledR1Fraction(
+            long numControls,
+            Qubit controls[],
+            Qubit target,
+            long numerator,
+            int power) = 0;
+        virtual void ControlledExp(
+            long numControls,
+            Qubit controls[],
+            long numTargets,
+            PauliId paulis[],
+            Qubit targets[],
+            double theta) = 0;
+        virtual void ControlledExpFraction(
+            long numControls,
+            Qubit controls[],
+            long numTargets,
+            PauliId paulis[],
+            Qubit targets[],
+            long numerator,
+            int power) = 0;
+
+        // Adjoint operations
+        virtual void SAdjoint(Qubit target) = 0;
+        virtual void TAdjoint(Qubit target) = 0;

```

Comment:
Why only these? Why not the adjoint for other Cliffords or for rotations?

Reply:
Full state simulator provides these as part of its public interface. I presume, this is because other operators are either self-adjoint and the adjoint can be created by modifying the arguments. Given the interface is temporary, I'll keep these for now to make sure all QIR can run against the full state simulator. In the future, the automatically generated base class will define the gates as it sees fit.